### ISSUE TITLE
[request] logind to listen for and respond to a rescue keyboard command sequence for hung sessions

### ISSUE THREAD

**Issue by n3rdopolis** [systemd/systemd#26313]
### Component

systemd-logind

### Is your feature request related to a problem? Please describe

Since logind already responds to power buttons, which are /dev/input/event(x) devices, just like normal keyboards, logind watching for keystrokes to respond with an action on the devices seat seems to be the best place to watch for more keystrokes.

As the possibility of VT-less world grows closer, the concern of being able to start a minimal shell when something goes irresponsive remains. And these issues might even already be possible today even with VTs, since it's the display server that handles CTRL+ALT+F(x), so if it's a full hang of the display server, it won't respond to CTRL+ALT+F(x). 

### Describe the solution you'd like

What the key stroke sequence will actually be is debatable. It should be something hard to do accidentally, and won't conflict with known bindings, but not too hard, or inaccessible, and if possible, should involve keys that not even over-minimalisticly oriented laptop manufacturers haven't randomly decided that said key is not "cool enough". Maybe CTRL+ALT+(PrtSc/SysRq) 3 times within one second.

The action that should actually happen is probably debatable too:

1.It could switch to an empty VT, but one where a getty is running on, or for seats that don't support VTs, on the seat of which the keyboard where the key stroke was pressed belongs to, start a user mode Wayland server that a terminal emulator runs a getty under. 
Although nothing like that is actually packaged yet, and there would probably be debate about the terminal emulator, and display server to actually use. 

2. It could switch to the display manager's greeter session on the seat. Although not every display manager keeps its greeter running, and I don't think there is a universal way to tell every display manager that it's time to show its greeter. Also, some start their greeter display server as root, so they are not logind sessions (sddm). Unless if maybe it would be something like `session-rescue@{seat}.target` and then the display manager would have to ship services that make it WantedBy the target?

3. It could kill the session on the seat, probably the easiest, it would be smarter than ALT+SysRq+K even, but users might still find that destructive in some cases, like if the user needs to kill a single process using a ton of CPU or something.

### The systemd version you checked that didn't have the feature you are asking for

253

**Comment by n3rdopolis** [comment]
Perhaps there can be something like display-manager-rescue@.service that the Display Manager will have to create (where the parameter is the originating seat from the keyboard)

when a command like CTRL+ALT+PrintSc or something is pressed something like 3 times in 3 seconds or something, that service (perhaps a simple service) can be started. The expectation would be that the command that gets run tells the Display Manager to either switch to, or start its greeter?

**Comment by n3rdopolis** [comment]
Since logind can detect long holds of button/key presses, maybe pressing and holding CTRL+SHIFT+ESC  can get logind to emit a dbus signal like 
```
SeatWantsGreeter(s seat_id,
                       o object_path);
```
for display managers to listen for. Does this make sense? Or do developers of display managers need to weigh in?

**Comment by n3rdopolis** [comment]
I am trying to experiment, The issue is that it seems that I will have to make CONST_MAX5 into CONST_MAX14 I think right, adding ESC, as well as both the CTRL, ALT, Meta, Shift keys, which seems sort of a lot of nested CONST_MAX's

I am trying to make the sequence pressing and holding CTRL+SHIFT+ESC, but if I allow CTRL and SHIFT through `EVIOCSMASK` that would make logind get woken up when those keys are hit, right

I have a local diff, but I am not sure why the attaching function isn't working right now.

**Comment by n3rdopolis** [comment]
This is what I have so far. I am struggling with making the CONST_MAX5 into CONST_MAX14 [sd.diff.txt](https://github.com/systemd/systemd/files/12434615/sd.diff.txt)

and I am now second guessing it as logind would get waken up each time a shift, ctrl, alt, or meta key is pressed.

Maybe it should be Caps+Esc? It has to be something that is not hit often, but it has to be something on the majority of keyboards, including extremely simplified ones in those mac books with no function keys.



Press Caps+Esc, and then emit some sort of dbus signal on the seat of the device is what I am aiming for

**Comment by n3rdopolis** [comment]
Fixed by #29542
